tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Lambda/HandlerWrapperSetHandlerIntegration.cs

```diff
@@ -0,0 +1,139 @@
+    /// <returns>Calltarget state value</returns>
+    internal static CallTargetState OnMethodBegin<TTarget>(TTarget instance, ref Delegate handler)
+    {
+        handler = handler.Instrument(_callbacks);
+        return CallTargetState.GetDefault();
+    }
+
+    private static string ConvertPayloadStream(Stream payloadStream)
+    {
+        var reader = new StreamReader(payloadStream, Encoding.UTF8, leaveOpen: true);
+        var result = reader.ReadToEnd();
+        payloadStream.Seek(0, SeekOrigin.Begin); // Reset the offset so that it can be read by the originally intended consumer
+        return result;
+    }
+
+    private readonly struct Async1Callbacks : IBegin1Callbacks, IReturnCallback, IReturnAsyncCallback
+    {
+        public bool PreserveAsyncContext => false;
+
+        public object OnDelegateBegin<TArg1>(object sender, ref TArg1 arg)
+        {
+            Serverless.Debug("DelegateWrapper Running OnDelegateBegin");
+            try
+            {
+                var proxyInstance = arg.DuckCast<IInvocationRequest>();
+                if (proxyInstance != null)
+                {
+                    var json = ConvertPayloadStream(proxyInstance.InputStream);
+                    var scope = LambdaCommon.SendStartInvocation(new LambdaRequestBuilder(), json, proxyInstance.LambdaContext?.ClientContext?.Custom);
+                    return new CallTargetState(scope);
+                }
+                else
+                {
+                    Serverless.Debug("DuckCast.IInvocationRequest got null proxyInstance");
+                    var scope = LambdaCommon.SendStartInvocation(new LambdaRequestBuilder(), string.Empty, null);
+                    return new CallTargetState(scope);
+                }
+            }
+            catch (Exception ex)
+            {
+                Serverless.Error("Could not send payload to the extension", ex);
+                Console.WriteLine(ex.StackTrace);
+            }
+
+            Serverless.Debug("DelegateWrapper FINISHED Running OnDelegateBegin");
+            return CallTargetState.GetDefault();
+        }
+
+        /// <inheritdoc/>
+        public TReturn OnDelegateEnd<TReturn>(object sender, TReturn returnValue, Exception exception, object state)
+        {
+            return returnValue;
+        }
+
+        /// <inheritdoc/>
+        public Task<TInnerReturn> OnDelegateEndAsync<TInnerReturn>(object sender, TInnerReturn returnValue, Exception exception, object state)
+        {
+            Serverless.Debug("DelegateWrapper Running OnDelegateEndAsync");
+            try
+            {
+                var proxyInstance = returnValue.DuckCast<IInvocationResponse>();
+                if (proxyInstance != null)
+                {
+                    var json = ConvertPayloadStream(proxyInstance.OutputStream);
+                    LambdaCommon.EndInvocationSync(json, exception, ((CallTargetState)state!).Scope, RequestBuilder);
```

Comment:
Given this is an async method I think we could use `LambdaCommon.EndInvocationAsync`? If so that's probably preferable to avoid any potential sync-over-async issues

Reply:
Updated while cleaning up.